### PR TITLE
fix: name temporary variables with namespace

### LINKED ISSUE #777: Variable declaration pollutes other variables
**test.xsl** defines a variable, `$my:test-doc`:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet exclude-result-prefixes="#all" version="2.0" xmlns:my="my-namespace"
	xmlns:xsl="http://www.w3.org/1999/XSL/Transform">

	<xsl:variable as="document-node()" name="my:test-doc">
		<xsl:document>
			<test>document defined in SUT</test>
		</xsl:document>
	</xsl:variable>

</xsl:stylesheet>
```

**test.xspec** defines `$my:test` and expects `$my:test-doc` to be intact:
```xspec
<?xml version="1.0" encoding="UTF-8"?>
<x:description stylesheet="test.xsl" xmlns:my="my-namespace"
	xmlns:x="http://www.jenitennison.com/xslt/xspec">

	<x:param as="element(test)" name="my:test">
		<test>element defined in XSpec</test>
	</x:param>

	<x:scenario label="When x:variable in SUT defines $my:test-doc">
		<x:call function="false" />
		<x:expect label="$my:test-doc should be an document defined in SUT" select="/"
			test="$my:test-doc">
			<test>document defined in SUT</test>
		</x:expect>
	</x:scenario>

</x:description>
```

This `x:expect` fails:
```console
M:\test>\xspec\bin\xspec.bat test.xspec
...
Running Tests...
Testing with SAXON EE 9.9.1.6
When x:variable in SUT defines $my:test-doc
$my:test-doc should be an document defined in SUT
      FAILED
```

[`test:generate-variable-declarations`](https://github.com/xspec/xspec/blob/c37bdb30c27ee431854adb051c8ed460855e5e69/src/compiler/generate-tests-helper.xsl#L38-L121) overwrites `$my:test-doc` with a document node whose content is
```xml
<test xmlns:my="my-namespace">element defined in XSpec</test>
```

### PR DESCRIPTION
**Pull request by AirQuick**
Fixes #777 

Before fixing the problem, the temporary variable names were `[final name]-doc` and `[final name]-doc-uri`. They can conflict with user-defined variable names.

This pull request changes the naming convention: `[private namespace prefix]:[context element local name]-[context element id]-doc` and `...-uri`.
`[private namespace prefix]` is `local` (XQuery) or [`impl`](https://github.com/xspec/xspec/blob/36920399852e3ecfe7c2a215c9cd6e198bba19e9/src/compiler/generate-xspec-tests.xsl#L54) (XSLT).

### CODE REVIEW

**Comment by AirQuick** [comment]
@galtm 
Before making this change, I refactored the relevant code (in #782 and #792). That's why many lines are different from v1.6.0. I believe the refactoring has made it easier to review this.

File: test/xspec-777.xspec [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<x:description query="my-namespace" query-at="xspec-777.xquery" stylesheet="xspec-777.xsl"
+	xmlns:my="my-namespace" xmlns:x="http://www.jenitennison.com/xslt/xspec">
+
+	<x:param as="document-node(element(test))" href="xspec-777.xml" name="my:test" />
+
+	<x:scenario>
+		<x:label>When x:param[@href] in XSpec defines $my:test and x:variable (or its XQuery
```
Comment (galtm) [inline test/xspec-777.xspec]: Do you mean `xsl:variable` instead of `x:variable`?

**Comment by galtm** [comment]
> Before making this change, I refactored the relevant code (in #782 and #792). That's why many lines are different from v1.5.0. I believe the refactoring has made it easier to review this.

Thanks, @AirQuick . I've started looking at this. I need a little more time that's not at the end of a long day.

File: test/xspec-777.xspec (hunk 1 above)
Comment (AirQuick) [reply test/xspec-777.xspec]: Fixed in f3dcdde310da818d355535c9cd8a1fdfddeff7bb

**Comment by AirQuick** [comment]
Thanks for the reivew, @galtm 

> The [Compilation wiki](https://github.com/xspec/xspec/blob/master/tutorial/under-the-hood/Compilation.md) might need some updates. No hurry. If you're too busy with other pull requests and you'd like me to update the wiki, let me know.

I forgot about it! although we have a reminder (#505).
Can you try updating it?

A problem is that there is no `.xspec` file (apart from the variable files you created) that is connected to the description in `Compilation.md`. To facilitate updating `Compilation.md`, it would be nice if we have an actual `.xspec` file.

**Comment by galtm** [comment]
> I forgot about it! although we have a reminder (#505).
> Can you try updating it?
> 
> A problem is that there is no `.xspec` file (apart from the variable files you created) that is connected to the description in `Compilation.md`. To facilitate updating `Compilation.md`, it would be nice if we have an actual `.xspec` file.

I forgot about #505 . I'll assign it to myself. Do you think it's ready to be worked on anytime? The description says (from over a year ago), "At some point in the future when the outstanding pull requests are resolved".

**Comment by AirQuick** [comment]
Soon I'm going to open a small refactoring pull request to get ready for [dynamic invocation](https://github.com/xspec/xspec/issues/762#issuecomment-593150373). And I was going to ask you to review it. It's small, but it may change something in the compiled stylesheet. I'll let you know.

Another idea is that the compiler itself should possibly write comments like [`<!-- the main template to run the suite -->`](https://github.com/xspec/xspec/blame/master/tutorial/under-the-hood/Compilation.md#L66) into the compiled stylesheet/query in the first place. Just my quick thought.

**Comment by galtm** [comment]
OK, I will wait for (at least) your refactoring pull request before working on the compilation wiki. I can review that PR when it's ready.

I tend to lean toward seeing code comments as useful rather than clutter, so I like your idea about making the compiler insert some of the comments that are currently added manually to the compilation wiki.

**Comment by AirQuick** [comment]
> OK, I will wait for (at least) your refactoring pull request before working on the compilation wiki. I can review that PR when it's ready.

Opened #807. Fortunately, I believe nothing has changed in the compiled stylesheet.

**Pull request merged**